text/0000-template-invocation-context.md

```diff
@@ -0,0 +1,156 @@
+// functional modifier
+// class-based helper
+
+let templateStack = named._getInvocationStack();
+
+// class-based modifier
+
+let templateStack = this.args.named._getInvocationStack();
+
+// component
+
+let templateStack = this.args._getInvocationStack();
+```
+
+In production, this method could return an empty object.
+
+a') Same as (a) but with a property getter instead of a function reference: e.g. `let templateStack = named._invocationStack;`
+
+b) A new debug function that can extract the stack from a hidden (symbol?) property stashed on the named/args object. It would look like this:
+
+```js
+import { getTemplateLocationInformation } from '@glimmer/debug';
+
+// functional helper
+// functional modifier
+// class-based helper
+
+let templateStack = getTemplateLocationInformation(named);
+
+// class-based modifier
+
+let templateStack = getTemplateLocationInformation(this.args.named);
+
+// component
+
+let templateStack = getTemplateLocationInformation(this.args);
+```
+
+In production, this method could return an empty object.
+
+In either case, we could consider a lint rule to make sure this is invoked within an `if (DEBUG) { ... }` block.
+
+### 3) How is the stack information is tracked by Glimmer and collected when requested via this API?
+
+I need help on this section. From discussing briefly with @rwjblue, I have learned that one of the precursors for implementation of something like this is to ensure that Ember's internal template compilation is aware of debug vs prod builds (today we compile templates with the same set of built-in AST plugins). This would be beneficial to reducing production build size, because the lack of differentiation in the current system has caused template size bloat because the compiler must always include debug information.
+
+## How we teach this
+
+This should be included in the API docs for helpers, modifiers and components with example code shown, along with example output. If a configuration option is included like the one posited under "Alternatives," this should be included in the guides under "Configuration > Debugging".
+
+When released, it should be included in the Ember.js Blog post for the release as well as highlighted in Ember Times. This could save existing Ember developers a lot of time and frustration.
+
+## Drawbacks
+
+APIs for helpers and modifiers are currently wonderfully simple and elegant. The addition this ugly debugging API compromises that design.
+
+The fact that that this will only be available in debug builds adds to complexity for consumers of the API and could potentially lead to errors in production if a developer does not realize this.
+
+## Alternatives
+
+@rwjblue created an addon to address this problem: [ember-template-invocation-location](https://github.com/rwjblue/ember-template-invocation-location). This work is the basis for much of the content of this RFC. One alternative is to point the community toward this addon and not make any changes to core.
```

Comment:
IMHO, this addon isn’t quite good enough. It requires each “layer” to have a dependency on it for it to function through the addon template layer, and that basically means that you are highly unlikely to consistently ensure your stack leads all the way back to the root.